File: 2025/day09.cpp
// https://adventofcode.com/2025/day/9

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>
#include <ranges>

using Point = aoc::point_type<i64>;
using Segment = aoc::closed_range<Point>;
using Polygon = Vec<Point>;
using Rectangle = std::array<Point, 4>;

auto parse(String const& filename) -> Polygon {
  return aoc::views::read_lines(filename) |
         stdv::transform(
             [](str line) { return aoc::split<Point>(line, ','); }) |
         aoc::ranges::to<Polygon>();
}

fn area(Point const& p1, Point const& p2) -> u64 {
  let diff = (p2 - p1).abs() + Point{1, 1};
  return static_cast<u64>(diff.x) * static_cast<u64>(diff.y);
}

fn solve_case1(Polygon const& polygon) -> u64 {
  return stdr::max( //
      stdv::cartesian_product(polygon, polygon) |
      stdv::filter([](auto&& pair) {
        let & [ p1, p2 ] = pair;
        return p1 < p2;
      }) |
      stdv::transform([](auto&& pair) {
        let & [ p1, p2 ] = pair;
        return area(p1, p2);
      }));
}

fn point_on_segment(Point const& point, Segment const& segment) -> bool {
  return aoc::orientation(segment.begin, segment.end, point) == 0 &&
         segment.contains(point);
}

fn point_inside_polygon(Polygon const& polygon, Point const& point) -> bool {
  let vertex_count = polygon.size();
  auto inside = false;
  auto prev_idx = vertex_count - 1;
  // Plain index loops rather than Range: these two are the innermost loops of
  // part 2, run once per polygon vertex per candidate rectangle
  for (auto curr_idx = 0uz; curr_idx < vertex_count; ++curr_idx) {
    let& curr_vertex = polygon[curr_idx];
    let& prev_vertex = polygon[prev_idx];
    if (point_on_segment(point, Segment{curr_vertex, prev_vertex})) {
      // The boundary is inside
      return true;
    }
    // Handle vertices on the ray without double-counting
    if ((curr_vertex.y > point.y) != (prev_vertex.y > point.y)) {
      let delta_y = prev_vertex.y - curr_vertex.y;
      let lhs = (point.x - curr_vertex.x) * delta_y;
      let rhs = (prev_vertex.x - curr_vertex.x) * (point.y - curr_vertex.y);
      // Avoid integer division by rearranging the crossing comparison
      if ((lhs < rhs) == (delta_y > 0)) {
        inside = !inside;
      }
    }
    prev_idx = curr_idx;
  }
  return inside;
}

fn corners_inside(Polygon const& polygon, Rectangle const& rectangle) -> bool {
  return stdr::all_of(rectangle, [&](Point const& corner) {
    return point_inside_polygon(polygon, corner);
  });
}

fn segments_cross(Segment const& seg1, Segment const& seg2) -> bool {
  let o1 = aoc::orientation(seg2.begin, seg2.end, seg1.begin);
  let o2 = aoc::orientation(seg2.begin, seg2.end, seg1.end);
  // Collinear: shared boundary segment, not a proper intersection
  if (o1 == 0 && o2 == 0) {
    return false;
  }
  let o3 = aoc::orientation(seg1.begin, seg1.end, seg2.begin);
  let o4 = aoc::orientation(seg1.begin, seg1.end, seg2.end);
  // Proper crossing: each segment strictly straddles the other's line
  return (o1 * o2 < 0) && (o3 * o4 < 0);
}

fn edges_intersect(Polygon const& polygon, Rectangle const& rectangle) -> bool {
  // rectangle indices:
  //   0=(xmin,ymin),
  //   1=(xmin,ymax),
  //   2=(xmax,ymin),
  //   3=(xmax,ymax)
  let rect_edges = std::array{
      Segment{rectangle[0], rectangle[1]}, // left
      Segment{rectangle[2], rectangle[3]}, // right
      Segment{rectangle[0], rectangle[2]}, // bottom
      Segment{rectangle[1], rectangle[3]}, // top
  };
  let edge_count = polygon.size();
  auto prev_idx = edge_count - 1;
  for (auto curr_idx = 0uz; curr_idx < edge_count; ++curr_idx) {
    let poly_segment = Segment{polygon[curr_idx], polygon[prev_idx]};
    for (let& rect_segment : rect_edges) {
      if (segments_cross(poly_segment, rect_segment)) {
        return true;
      }
    }
    prev_idx = curr_idx;
  }
  return false;
}

// A pair of corners with its area worked out up front
struct SizedPair {
  u64 area;
  Point p1;
  Point p2;
};

fn solve_case2(Polygon const& polygon) -> u64 {
  // The area is stored alongside each pair rather than recomputed inside the
  // comparator, which called it twice for every one of the ~2M comparisons.
  // Plain loops rather than cartesian_product, for the usual Debug reasons.
  auto pairs = Vec<SizedPair>{};
  for (let& p1 : polygon) {
    for (let& p2 : polygon) {
      if (p1 < p2) {
        pairs.emplace_back(area(p1, p2), p1, p2);
      }
    }
  }
  stdr::sort(pairs, [](SizedPair const& a, SizedPair const& b) {
    return a.area > b.area;
  });
  for (let& [ pair_area, p1, p2 ] : pairs) {
    let xmin = std::min(p1.x, p2.x);
    let xmax = std::max(p1.x, p2.x);
    let ymin = std::min(p1.y, p2.y);
    let ymax = std::max(p1.y, p2.y);
    let corners = Rectangle{
        Point{xmin, ymin},
        Point{xmin, ymax},
        Point{xmax, ymin},
        Point{xmax, ymax},
    };
    if (corners_inside(polygon, corners) &&
        !edges_intersect(polygon, corners)) {
      return pair_area;
    }
  }
  return 0;
}

int main() {
  std::println("Part 1");
  let example = parse("day09.example");
  AOC_EXPECT_RESULT(50, solve_case1(example));
  let input = parse("day09.input");
  AOC_EXPECT_RESULT(4744899849, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(24, solve_case2(example));
  AOC_EXPECT_RESULT(1540192500, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2025/src/day09.rs
use itertools::Itertools;

type Point = aoc::point::Point<i64>;
type Polygon = Vec<Point>;

fn parse(filename: &str) -> Polygon {
    let parse = |s: &str| s.parse::<i64>().unwrap();
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let coords = line.split(',').collect::<Vec<_>>();
            Point::new(parse(coords[0]), parse(coords[1]))
        })
        .collect()
}

fn area(p1: &Point, p2: &Point) -> u64 {
    let diff = (*p2 - *p1).abs() + Point::new(1, 1);
    (diff.x as u64) * (diff.y as u64)
}

fn solve_case1(polygon: &Polygon) -> u64 {
    polygon
        .iter()
        .cartesian_product(polygon.iter())
        .filter(|(p1, p2)| p1 < p2)
        .map(|(p1, p2)| area(p1, p2))
        .max()
        .unwrap()
}

type Rectangle = [aoc::point::Point<i64>; 4];

fn point_on_segment(point: &Point, seg_start: &Point, seg_end: &Point) -> bool {
    // Check collinearity via cross product
    let cross = (seg_end.x - seg_start.x) * (point.y - seg_start.y)
        - (seg_end.y - seg_start.y) * (point.x - seg_start.x);
    // Bounds check coordinates
    cross == 0
        && point.x >= seg_start.x.min(seg_end.x)
        && point.x <= seg_start.x.max(seg_end.x)
        && point.y >= seg_start.y.min(seg_end.y)
        && point.y <= seg_start.y.max(seg_end.y)
}

fn point_inside_polygon(polygon: &Polygon, point: &Point) -> bool {
    let vertex_count = polygon.len();
    let mut inside = false;
    let mut prev_idx = vertex_count - 1;
    for curr_idx in 0..vertex_count {
        let (curr_vertex, prev_vertex) = (&polygon[curr_idx], &polygon[prev_idx]);
        if point_on_segment(point, curr_vertex, prev_vertex) {
            // The boundary is inside
            return true;
        }
        // Handle vertices on the ray without double-counting
        if (curr_vertex.y > point.y) != (prev_vertex.y > point.y) {
            let delta_y = prev_vertex.y - curr_vertex.y;
            let lhs = (point.x - curr_vertex.x) * delta_y;
            let rhs = (prev_vertex.x - curr_vertex.x) * (point.y - curr_vertex.y);
            // Avoid integer division by rearranging the crossing comparison
            if (lhs < rhs) == (delta_y > 0) {
                inside = !inside;
            }
        }
        prev_idx = curr_idx;
    }
    return inside;
}

/// Run a standard ray casting test for each corner.
/// Treat polygon on an edge as inside.
/// If any corner is outside, return false.
fn corners_inside(polygon: &Polygon, rectangle: &Rectangle) -> bool {
    rectangle
        .iter()
        .all(|corner| point_inside_polygon(polygon, corner))
}

fn cross2d(from: &Point, to: &Point, point: &Point) -> i64 {
    (to.x - from.x) * (point.y - from.y) - (to.y - from.y) * (point.x - from.x)
}

fn segments_cross(
    seg1_start: &Point,
    seg1_end: &Point,
    seg2_start: &Point,
    seg2_end: &Point,
) -> bool {
    let seg1_start_side = cross2d(seg2_start, seg2_end, seg1_start);
    let seg1_end_side = cross2d(seg2_start, seg2_end, seg1_end);
    let seg2_start_side = cross2d(seg1_start, seg1_end, seg2_start);
    let seg2_end_side = cross2d(seg1_start, seg1_end, seg2_end);
    // Collinear: shared boundary segment, not a proper intersection
    if seg1_start_side == 0 && seg1_end_side == 0 {
        return false;
    }
    // Proper crossing: each segment strictly straddles the other's line
    return (seg1_start_side > 0 && seg1_end_side < 0 || seg1_start_side < 0 && seg1_end_side > 0)
        && (seg2_start_side > 0 && seg2_end_side < 0 || seg2_start_side < 0 && seg2_end_side > 0);
}

/// For each polygon edge check whether it properly
/// intersects any rectangle edge.
/// If any intersection exists that is not just a shared boundary segment,
/// returns false.
fn edges_intersect(polygon: &Polygon, rectangle: &Rectangle) -> bool {
    // rectangle indices: 0=(xmin,ymin), 1=(xmin,ymax), 2=(xmax,ymin), 3=(xmax,ymax)
    let rect_edges: [(&Point, &Point); 4] = [
        (&rectangle[0], &rectangle[1]), // left
        (&rectangle[2], &rectangle[3]), // right
        (&rectangle[0], &rectangle[2]), // bottom
        (&rectangle[1], &rectangle[3]), // top
    ];
    let edge_count = polygon.len();
    let mut prev_idx = edge_count - 1;
    for curr_idx in 0..edge_count {
        let (poly_start, poly_end) = (&polygon[curr_idx], &polygon[prev_idx]);
        for &(rect_start, rect_end) in &rect_edges {
            if segments_cross(poly_start, poly_end, rect_start, rect_end) {
                return true;
            }
        }
        prev_idx = curr_idx;
    }
    return false;
}

fn solve_case2(polygon: &Polygon) -> u64 {
    polygon
        .iter()
        .cartesian_product(polygon.iter())
        .map(|(&p1, &p2)| {
            if (p1 >= p2) {
                return 0;
            }
            let (xmin, xmax) = (p1.x.min(p2.x), p1.x.max(p2.x));
            let (ymin, ymax) = (p1.y.min(p2.y), p1.y.max(p2.y));
            let corners = [
                Point::new(xmin, ymin),
                Point::new(xmin, ymax),
                Point::new(xmax, ymin),
                Point::new(xmax, ymax),
            ];
            if (!corners_inside(polygon, &corners) || edges_intersect(polygon, &corners)) {
                return 0;
            }
            return area(&p1, &p2);
        })
        .max()
        .unwrap()
}

fn main() {
    println!("Part 1");
    let example = parse("day09.example");
    aoc::expect_result!(50, solve_case1(&example));
    let input = parse("day09.input");
    aoc::expect_result!(4744899849, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(24, solve_case2(&example));
    aoc::expect_result!(1540192500, solve_case2(&input));
}
